sqlliveness: embed current region into session ID

jaylim-crl opened the pull request:
Follow up on #90408 and #91019.

Previously, the sqlliveness subsystem was using enum.One when constructing the
session ID since it doesn't have the regional data yet. This commit implements
that missing part of it by ensuring that the regional representation is
plumbed all the way to the sqlliveness subsystem whenever the SQL pod is
started, and use it to construct the session ID. This will enable our REGIONAL
BY ROW work to put the data in the right region.

The multi-region enum for the system database will be read during startup for
that to work. Since doing this already requires loading the system DB's
descriptor (which indirectly tests whether the system DB has been bootstrapped)
for the tenant, we can remove the call to checkTenantExists.

Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/91694)
<!-- Reviewable:end -->

jaylim-crl reviewed (commented):
I would need to write a test for the DescsTxn work in server_sql.go, but the rest should be here. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91694#-:-NGYBu1PAJqzhvD9QsHJ:bo5z9oq)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

postamar commented on pkg/sql/sqlliveness/sqlliveness.go:
```diff
@@ -20,12 +20,23 @@ import (
 
+// TypeDescriptor represents a stripped down version of catalog.TypeDescriptor.
+// We do this to avoid a circular dependency on pkg/sql/catalog, which imports
+// pkg/sql/sem/eval. The latter package imports pkg/sql/sqlliveness.
```
This dependency of catalog on eval is a shameful wart which the SQL Schema team has meant to remove for a while now but still hasn't. Sorry about that!

In any case even if that obstacle weren't in the way, the dependency structure of your change bothers me a bit. I may not have the full picture here but it feels like you're pushing too much high-level logic into `SetRegionalData` when in fact it would be better off being more "dumb" and instead just take the `[]byte`s for the physical representation of the region, either nakedly or wrapped in some lightweight abstraction.

This could in turn originate from a new method you could add on `multiregion.RegionConfig` and which you call in the `server` package. This way no need to shadow `catalog.TypeDescriptor` at all, which is just as well.

postamar commented on pkg/server/server_sql.go:
```diff
@@ -1306,13 +1284,44 @@ func (s *SQLServer) preStart(
+		return err
+	}); err != nil {
+		return err
+	}
```
We have a `SynthesizeRegionConfig` function in `sql` which does most of this already, would it be possible to use it? You'd have to extend `multiregion.RegionConfig` to include the physical representation, presumably, but that seems reasonable.

postamar reviewed (commented):
I just have a couple of high-level comments here, which may be wrong. @ajwerner will no doubt to look at this more closely and have stronger opinions than me :)

ajwerner reviewed (commented):
> Reviewed 5 of 7 files at r1, 6 of 6 files at r4, 3 of 3 files at r5, 3 of 3 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91694)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @JeffSwenson and @postamar)

___
*[`pkg/server/server_sql.go` line 1320 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/91694#-NGY297YSadImbgOrfE8-r7-1320:-NGr3nck0X_CwqjOiy3U:bj1yeay) ([raw file](https://github.com/cockroachdb/cockroach/blob/d900b6bb8795d48e5af78bacf7ffb678879448c9/pkg/server/server_sql.go#L1320)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

We have a `SynthesizeRegionConfig` function in `sql` which does most of this already, would it be possible to use it? You'd have to extend `multiregion.RegionConfig` to include the physical representation, presumably, but that seems reasonable.
</blockquote></details>

My vote is that we share the parts that are used also by `SynthesizeRegionConfig` and make this a function in `regionutil` also. 
___
*[`pkg/sql/sqlliveness/sqlliveness.go` line 34 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/91694#-NGY297YSadImbgOrfE4-r7-34:-NGr31UP7vtTYNW226VW:b6c4ljk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d900b6bb8795d48e5af78bacf7ffb678879448c9/pkg/sql/sqlliveness/sqlliveness.go#L34)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This dependency of catalog on eval is a shameful wart which the SQL Schema team has meant to remove for a while now but still hasn't. Sorry about that!

In any case even if that obstacle weren't in the way, the dependency structure of your change bothers me a bit. I may not have the full picture here but it feels like you're pushing too much high-level logic into `SetRegionalData` when in fact it would be better off being more "dumb" and instead just take the `[]byte`s for the physical representation of the region, either nakedly or wrapped in some lightweight abstraction.

This could in turn originate from a new method you could add on `multiregion.RegionConfig` and which you call in the `server` package. This way no need to shadow `catalog.TypeDescriptor` at all, which is just as well.
</blockquote></details>

FWIW, I think this narrowing of an interface is totally fine. I'm not sure I'm in love with falling back to the `*descpb.TypeDescriptor`. I think you could narrow it even further if you wanted. My 2c is you could just as well pass a struct that has the primary region, the process's locality, and the mapping of region to prefix bytes? 


<!-- Sent from Reviewable.io -->

postamar commented:
>  I'm not sure I'm in love with falling back to the *descpb.TypeDescriptor

I'd go stronger in that we definitely should not do that! If we want to go further on this path of defining a `sqlliveness.TypeDescriptor` shadow-interface, the thing to do would be: remove the `TypeDesc` method and add `RegionNames`, `NumEnumMembers`, `GetMemberPhysicalRepresentation` and `GetMemberLogicalRepresentation` and implement `SetRegionalData` using those. It's a bit contrived, but the type descriptor protobuf has no place there.

jaylim-crl committed:
sqlliveness: embed current region into session ID

Follow up on #90408 and #91019.

Previously, the sqlliveness subsystem was using enum.One when constructing the
session ID since it doesn't have the regional data yet. This commit implements
that missing part of it by ensuring that the regional representation is
plumbed all the way to the sqlliveness subsystem whenever the SQL pod is
started, and use it to construct the session ID. This will enable our REGIONAL
BY ROW work to put the data in the right region.

The multi-region enum for the system database will be read during startup for
that to work. Since doing this already requires loading the system DB's
descriptor (which indirectly tests whether the system DB has been bootstrapped)
for the tenant, we can remove the call to checkTenantExists.

Epic: None

Release note: None

jaylim-crl reviewed (commented):
TFTRs! I cleaned up sqlliveness to only take in a `[]byte` slice so no additional dependencies are introduced. The logic of choosing the current region was pushed to the caller as well.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91694#-:-NGsIXcV6fOs-DIVBCr6:b-knscw3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @JeffSwenson, and @postamar)

___
*[`pkg/server/server_sql.go` line 1320 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/91694#-NGY297YSadImbgOrfE8-r7-1320:-NGsI4tcDr8TB22h1Ews:br89udz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d900b6bb8795d48e5af78bacf7ffb678879448c9/pkg/server/server_sql.go#L1320)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

My vote is that we share the parts that are used also by `SynthesizeRegionConfig` and make this a function in `regionutil` also. 
</blockquote></details>

Done. Moved all the main logic to `GetRegionEnumRepresentations`. Did not extend `multiregion.RegionConfig` since we have multiple callers that were constructing regional configs, and don't have the physical representations available.
___
*[`pkg/sql/region_util.go` line 1554 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/91694#-NGsI_w85TksYJjFsRoC:-NGsI_w90Ju79WH4OtQB:bd23tx5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b79d1556a32a2df46cc159f59a0e388205b33e54/pkg/sql/region_util.go#L1554)):*
> ```Go
> 			CommonLookupFlags: tree.CommonLookupFlags{
> 				AvoidLeased:    !useCache,
> 				Required:       true,
> ```

Note that one additional change here is to specify Required=true when looking up the multi-region enum descriptor, which wasn't done previously. I'd imagine this should be OK.
___
*[`pkg/sql/sqlliveness/sqlliveness.go` line 50 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91694#-NGY297YSadImbgOrfE4-r1-50:-NGsIO6ABP_qUSKeGLcu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d139d457f96bb9eb20a534dec88bea71676f49e0/pkg/sql/sqlliveness/sqlliveness.go#L50)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

nit: Instead of introducing the SetRegionalData method, the locality information should be an argument to Start. SetRegionalData must be called before Start and the easiest way to ensure that is to fuse the methods into one.
</blockquote></details>

Done.
___
*[`pkg/sql/sqlliveness/sqlliveness.go` line 34 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/91694#-NGY297YSadImbgOrfE4-r7-34:-NGsIP4z47GvS9T_nzfN:b2gn3w1) ([raw file](https://github.com/cockroachdb/cockroach/blob/d900b6bb8795d48e5af78bacf7ffb678879448c9/pkg/sql/sqlliveness/sqlliveness.go#L34)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

FWIW, I think this narrowing of an interface is totally fine. I'm not sure I'm in love with falling back to the `*descpb.TypeDescriptor`. I think you could narrow it even further if you wanted. My 2c is you could just as well pass a struct that has the primary region, the process's locality, and the mapping of region to prefix bytes? 
</blockquote></details>

Done. Updated it to take in a `[]byte` instead so no additional dependencies are required.


<!-- Sent from Reviewable.io -->

jaylim-crl commented on pkg/sql/sqlliveness/slinstance/slinstance.go:
```diff
@@ -396,9 +401,12 @@ func NewSQLInstance(
-func (l *Instance) Start(ctx context.Context) {
+func (l *Instance) Start(ctx context.Context, regionPhysicalRep []byte) {
 	l.mu.Lock()
 	defer l.mu.Unlock()
```
There was a `started` bool back then, but has been removed in https://github.com/cockroachdb/cockroach/commit/e1d5835381da10afc371eadf5216a65963d6c602. We should be able to remove the mutex.

```go
// Start runs the hearbeat loop.
func (l *Instance) Start(ctx context.Context) {
	l.mu.Lock()
	defer l.mu.Unlock()
	if l.mu.started {
		return
	}
	log.Infof(ctx, "starting SQL liveness instance")
	_ = l.stopper.RunAsyncTask(ctx, "slinstance", l.heartbeatLoop)
	l.mu.started = true
}
```

jaylim-crl replied:
Will be removed in next push. Consider this done.

ajwerner reviewed (commented):
This is getting close. 

> Reviewed 1 of 2 files at r7, 11 of 12 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91694#-:-NGwVcI0-dpdobgptpXp:b-ix7odn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jaylim-crl, @JeffSwenson, and @postamar)

___
*[`pkg/sql/region_util.go` line 1554 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/91694#-NGsI_w85TksYJjFsRoC:-NGwTe0JDOzOGneNayeW:b-x9o3z6) ([raw file](https://github.com/cockroachdb/cockroach/blob/4073496004a1a6464f74a31d42d845b15141a1f5/pkg/sql/region_util.go#L1554)):*
<details><summary><i>Previously, jaylim-crl (Jay Lim) wrote…</i></summary><blockquote>

Note that one additional change here is to specify Required=true when looking up the multi-region enum descriptor, which wasn't done previously. I'd imagine this should be OK.
</blockquote></details>

yeah, for bad reasons, it was implied. https://github.com/cockroachdb/cockroach/blob/3038ea99ad457c53fa8ee418af6f94c92f2efd1d/pkg/sql/catalog/descs/type.go#L106-L107
___
*[`pkg/sql/sqlliveness/sqlliveness.go` line 37 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/91694#-NGwV-r92OeDbnEY_Mdi:-NGwV-r92OeDbnEY_Mdj:b-fqd6pe) ([raw file](https://github.com/cockroachdb/cockroach/blob/4073496004a1a6464f74a31d42d845b15141a1f5/pkg/sql/sqlliveness/sqlliveness.go#L37)):*
> ```Go
> // consumption.
> type Provider interface {
> 	Start(ctx context.Context, regionPhysicalRep []byte)
> ```

Please provide commentary explaining what this is. I even think you could define a type to make it read more cleanly, even if it's a type alias like like `type RegionEnumValue = []byte`
___
*[`pkg/sql/sqlliveness/slinstance/slinstance.go` line 406 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/91694#-NGY297YSadImbgOrfE1-r8-406:-NGwUwAj-hnfjrKAUNfN:b-2lszoc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4073496004a1a6464f74a31d42d845b15141a1f5/pkg/sql/sqlliveness/slinstance/slinstance.go#L406)):*
<details><summary><i>Previously, jaylim-crl (Jay Lim) wrote…</i></summary><blockquote>

There was a `started` bool back then, but has been removed in https://github.com/cockroachdb/cockroach/commit/e1d5835381da10afc371eadf5216a65963d6c602. We should be able to remove the mutex.

```go
// Start runs the hearbeat loop.
func (l *Instance) Start(ctx context.Context) {
	l.mu.Lock()
	defer l.mu.Unlock()
	if l.mu.started {
		return
	}
	log.Infof(ctx, "starting SQL liveness instance")
	_ = l.stopper.RunAsyncTask(ctx, "slinstance", l.heartbeatLoop)
	l.mu.started = true
}
```

</blockquote></details>

Yeah, I think the mutex is stale


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91694)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jaylim-crl, @JeffSwenson, and @postamar)

___
*[`pkg/sql/sqlliveness/slinstance/slinstance.go` line 406 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/91694#-NGY297YSadImbgOrfE1-r8-406:-NH6iqZoDRniITOCeyXx:b-uli6j8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4073496004a1a6464f74a31d42d845b15141a1f5/pkg/sql/sqlliveness/slinstance/slinstance.go#L406)):*
<details><summary><i>Previously, jaylim-crl (Jay Lim) wrote…</i></summary><blockquote>

Will be removed in next push. Consider this done.
</blockquote></details>

Did you forget to push?


<!-- Sent from Reviewable.io -->

jaylim-crl reviewed (commented):
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91694#-:-NHAcXKICH42hamLibxj:bz5mair)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @JeffSwenson, and @postamar)

___
*[`pkg/sql/region_util.go` line 1500 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/91694#-NGsI-1HtU4HOoR80mgx-r8-1500:-NHAcZ8TEsxT-G8xRMti:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4073496004a1a6464f74a31d42d845b15141a1f5/pkg/sql/region_util.go#L1500)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

nit: GetRegionEnumRepresentations is a good candidate for unit testing.
</blockquote></details>

Done.
___
*[`pkg/sql/region_util.go` line 1554 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/91694#-NGsI_w85TksYJjFsRoC:-NHAcZiwD8TEeK3D5kvm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4073496004a1a6464f74a31d42d845b15141a1f5/pkg/sql/region_util.go#L1554)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

yeah, for bad reasons, it was implied. https://github.com/cockroachdb/cockroach/blob/3038ea99ad457c53fa8ee418af6f94c92f2efd1d/pkg/sql/catalog/descs/type.go#L106-L107
</blockquote></details>

Done.
___
*[`pkg/sql/sqlliveness/sqlliveness.go` line 37 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/91694#-NGwV-r92OeDbnEY_Mdi:-NHAcaICAjfsShUtGtEa:b-b582m8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4073496004a1a6464f74a31d42d845b15141a1f5/pkg/sql/sqlliveness/sqlliveness.go#L37)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Please provide commentary explaining what this is. I even think you could define a type to make it read more cleanly, even if it's a type alias like like `type RegionEnumValue = []byte`
</blockquote></details>

Done. I left it as `[]byte` for now. Maybe when we start taking in the region physical representation in more places in the future, we can revisit this decision.
___
*[`pkg/sql/sqlliveness/slinstance/slinstance.go` line 406 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/91694#-NGY297YSadImbgOrfE1-r8-406:-NHAdEQQEjMzUr8W8-w7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4073496004a1a6464f74a31d42d845b15141a1f5/pkg/sql/sqlliveness/slinstance/slinstance.go#L406)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Did you forget to push?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jeffswenson reviewed (approved):
LGTM

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> :ship::it: 

> Reviewed 6 of 9 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91694#-:-NHW1_CV3-1vn3NVfcvy:b41aghk)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jaylim-crl, @JeffSwenson, and @postamar)



<!-- Sent from Reviewable.io -->

jaylim-crl commented on pkg/ccl/multiregionccl/multiregion_system_table_test.go:
```diff
@@ -234,3 +238,91 @@ func TestMrSystemDatabase(t *testing.T) {
+	})
+
+	// Update system database with regions.
+	tenSQLDB.Exec(t, `SET descriptor_validation = read_only`)
```
Done.

jaylim-crl commented on pkg/ccl/multiregionccl/multiregion_system_table_test.go:
```diff
@@ -234,3 +238,91 @@ func TestMrSystemDatabase(t *testing.T) {
+	tenSQLDB := sqlutils.MakeSQLRunner(tSQL)
+
+	// Set cluster setting override on the host, and wait for it to propagate.
+	sqlDB.Exec(t, fmt.Sprintf("ALTER TENANT $1 SET CLUSTER SETTING %s = 'true'",
```
Done.

jaylim-crl commented:
TFTRs!

bors r=JeffSwenson,ajwerner

jaylim-crl commented:
bors r=JeffSwenson,ajwerner
